### PR TITLE
Correlate volatile functions with enclosing FORs and in clauses

### PR DESCRIPTION
**Pull request by msullivan**
This means that things like `ORDER BY random()` and
`FILTER random() < 0.5` will work.

Part of #1784.

### CODE REVIEW

File: edb/pgsql/compiler/relgen.py [hunk 1]
```diff
@@ -1954,18 +1956,19 @@ def _compile_func_epilogue(
...
-        # Apply the volatility reference.
-        # See the comment in process_set_as_subquery().
-        func_rel.where_clause = astutils.extend_binop(
-            func_rel.where_clause,
-            pgast.NullTest(
-                arg=ctx.volatility_ref,
-                negated=True,
+    if expr.volatility is qltypes.Volatility.VOLATILE:
+        for ref in ctx.volatility_ref:
+            if isinstance(ref, context.NoVolatilitySentinel):
```
Comment (elprans) [inline edb/pgsql/compiler/relgen.py:1961]: I didn't like the `NO_VOLATILITY` hack very much before, but in this formulation I like it even less.  Perhaps we could track the `INSERT` subject `path_id` instead and condition computable correlation in `process_set_as_subquery` on that instead.

File: edb/pgsql/compiler/clauses.py [hunk 2]
```diff
@@ -77,6 +77,58 @@ def fini_stmt(
...
+        ctx: context.CompilerContextLevel) -> None:
+    if iterator is None:
+        return
+
+    old = () if is_cte else ctx.volatility_ref
+
+    path_id = iterator.path_id
+    ref: Optional[pgast.BaseExpr] = None
+
+    def get_ref() -> pgast.BaseExpr:
```
Comment (elprans) [inline edb/pgsql/compiler/clauses.py:123]: Can you explain why a callback approach is needed, please?

Comment (msullivan) [reply edb/pgsql/compiler/clauses.py:123]: I touched on it in the comment I added in `context.py` but could add a comment here too: it's so we don't add the volatility ref column to the relation unless it actually gets used.

File: edb/pgsql/compiler/clauses.py [hunk 3]
```diff
@@ -96,8 +150,12 @@ def compile_iterator_expr(
         # for path identity of the iterator expression.  This is
         # necessary to maintain correct correlation for the state
         # of iteration in DML statements.
-        relctx.ensure_bond_for_expr(
-            iterator_expr.expr.result, iterator_query, type='uuid', ctx=subctx)
+        # The already_existed check is to avoid adding in bogus volatility refs
+        # when we reprocess an iterator that was hoisted.
+        if not already_existed:
```
[the remark this answers is not in the document]
Comment (msullivan) [reply edb/pgsql/compiler/clauses.py:155]: I don't see how at a quick glance, at least

File: edb/pgsql/compiler/relgen.py (hunk 1 above)
Comment (msullivan) [reply edb/pgsql/compiler/relgen.py:1961]: Yeah, I think that should work.

**Comment by msullivan** [comment]
I think this should be good now?

**Review by elprans — APPROVED** [review]
Yep, looks good.

**Pull request merged**